pkg/sdkserver/sdkserver_test.go

```diff
@@ -678,176 +685,185 @@ func TestSDKServerUpdateEventHandler(t *testing.T) {
-			defer func() {
-				done <- struct{}{}
-			}()
+	sc, err := defaultSidecar(m)
 
-			// move past timeout
-			fc.Step(40 * time.Second)
-			return true, nil, k8serrors.NewConflict(schema.ParseGroupResource(""), "gameservers", errors.New("contention"))
-		})
+	assert.NoError(t, err)
+	stop := make(chan struct{})
+	sc.informerFactory.Start(stop)
+	assert.True(t, cache.WaitForCacheSync(stop, sc.gameServerSynced))
 
-		_, err := sc.Allocate(context.Background(), &sdk.Empty{})
-		assert.EqualError(t, err, "Allocation request timed out")
+	wg := sync.WaitGroup{}
+	wg.Add(1)
 
+	go func() {
+		err = sc.Run(stop)
+		assert.Nil(t, err)
+		wg.Done()
+	}()
+
+	assertStateChange := func(expected v1alpha1.GameServerState, timeout time.Duration, additional func(status v1alpha1.GameServerStatus)) {
 		select {
-		case <-done:
-		case <-time.After(5 * time.Second):
-			assert.Fail(t, "have contention")
+		case current := <-state:
+			assert.Equal(t, expected, current.State)
+			additional(current)
+		case <-time.After(timeout):
+			assert.Fail(t, "should have gone to Reserved by now")
 		}
-	})
+	}
+	noAdditional := func(status v1alpha1.GameServerStatus) {}
 
-	t.Run("unhealthy gameserver", func(t *testing.T) {
-		gs := defaultGs.DeepCopy()
-		gs.Status.State = v1alpha1.GameServerStateUnhealthy
-		_, sc, stop := setup(t, gs.DeepCopy())
-		defer close(stop)
+	// 3 seconds
+	_, err = sc.Reserve(context.Background(), &sdk.Duration{Seconds: 3})
+	assert.NoError(t, err)
 
-		_, err := sc.Allocate(context.Background(), &sdk.Empty{})
-		assert.EqualError(t, err, "cannot Allocate an Unhealthy GameServer")
+	assertStateChange(v1alpha1.GameServerStateReserved, 2*time.Second, func(status v1alpha1.GameServerStatus) {
+		assert.Equal(t, time.Now().Add(3*time.Second).Round(time.Second), status.ReservedUntil.Time.Round(time.Second))
 	})
 
-	t.Run("Shutdown gameserver", func(t *testing.T) {
-		gs := defaultGs.DeepCopy()
-		gs.Status.State = v1alpha1.GameServerStateShutdown
-		_, sc, stop := setup(t, gs.DeepCopy())
-		defer close(stop)
-
-		_, err := sc.Allocate(context.Background(), &sdk.Empty{})
-		assert.EqualError(t, err, "cannot Allocate a Shutdown GameServer")
+	// wait for the game server to go back to being Ready
+	assertStateChange(v1alpha1.GameServerStateRequestReady, 4*time.Second, func(status v1alpha1.GameServerStatus) {
+		assert.Nil(t, status.ReservedUntil)
 	})
 
-	t.Run("Deleting gameserver", func(t *testing.T) {
-		gs := defaultGs.DeepCopy()
-		now := metav1.Now()
-		gs.ObjectMeta.DeletionTimestamp = &now
-		_, sc, stop := setup(t, gs.DeepCopy())
-		defer close(stop)
+	// now test for being able to escape the Reserved state when requesting another state
+
+	// test ready
+	_, err = sc.Reserve(context.Background(), &sdk.Duration{Seconds: 3})
+	assert.NoError(t, err)
+	assertStateChange(v1alpha1.GameServerStateReserved, 2*time.Second, noAdditional)
```

Comment:
above the check after calling reserve had an assert, why not do one here too?

Reply:
Here you pass in the `noAdditional` function as the third argument. Above you did

```go
assertStateChange(v1alpha1.GameServerStateReserved, 2*time.Second, func(status v1alpha1.GameServerStatus) {
		assert.Equal(t, time.Now().Add(3*time.Second).Round(time.Second), status.ReservedUntil.Time.Round(time.Second))
})
```

So the first time you verify that the state changed via the callback but in subsequent times you don't.

Reply:
Oh! I see it now. I figured it might be redundant to test it again, but I see your point. I'll make the change!

Reply:
And done!